internal/antewrapper/msg_fees_decorator.go

```diff
@@ -0,0 +1,240 @@
+	}
+	return feeTx, nil
+}
+
+// EnsureSufficientMempoolFees verifies that the given transaction has supplied
+// enough fees(gas + additional fees) to cover a proposer's minimum fees. A result object is returned
+// indicating success or failure.
+//
+// Contract: This should only be called during CheckTx as it cannot be part of
+// consensus.
+func EnsureSufficientMempoolFees(ctx sdk.Context, gas uint64, feeCoins sdk.Coins, additionalFees sdk.Coins) error {
+	feeCoinsOriginal := feeCoins
+	requiredFees := sdk.Coins{}
+	minGasPrices := ctx.MinGasPrices()
+	if !minGasPrices.IsZero() {
+		requiredFees = make(sdk.Coins, len(minGasPrices))
+
+		// Determine the required fees by multiplying each required minimum gas
+		// price by the gas limit, where fee = ceil(minGasPrice * gasLimit).
+		glDec := sdk.NewDec(int64(gas))
+		for i, gp := range minGasPrices {
+			fee := gp.Amount.Mul(glDec)
+			requiredFees[i] = sdk.NewCoin(gp.Denom, fee.Ceil().RoundInt())
+		}
+	}
+
+	// Before checking gas prices, remove taxed from fee
+	var hasNeg bool
+	if feeCoins, hasNeg = feeCoins.SafeSub(additionalFees); hasNeg {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, DefaultInsufficientFeeMsg+": %q, required fees: %q = %q(base-fee) +%q(additional-fees)", feeCoins, requiredFees.Add(additionalFees...), requiredFees, additionalFees)
+	}
+
+	if !requiredFees.IsZero() && !feeCoins.IsAnyGTE(requiredFees) {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "Base Fee+additional fee cannot be paid with fee value passed in "+": %q, required: %q = %q(base-fee) +%q(additional-fees)", feeCoinsOriginal, requiredFees.Add(additionalFees...),
+			requiredFees, additionalFees)
+	}
+
+	return nil
+}
+
+func EnsureAccountHasSufficientFeesWithAcctBalanceCheck(gas uint64, feeCoins sdk.Coins, additionalFees sdk.Coins,
+	balancePerCoin sdk.Coins, minGasPriceForAdditionalFeeCalc sdk.Coin) error {
+	err := EnsureSufficientFees(gas, feeCoins, additionalFees, minGasPriceForAdditionalFeeCalc)
+	if err != nil {
+		return err
+	}
+
+	if balancePerCoin.IsZero() || !balancePerCoin.IsAnyGTE(feeCoins) {
+		return fmt.Errorf("fee payer account does not have enough balance to pay for %q", feeCoins)
+	}
+	return nil
+}
+
+// EnsureSufficientFees to be used by msg_service_router
+func EnsureSufficientFees(gas uint64, feeCoins sdk.Coins, additionalFees sdk.Coins,
+	minGasPriceForAdditionalFeeCalc sdk.Coin) error {
+	// Step 1. Check if fees has enough money to pay additional fees.
+	var hasNeg bool
+	if feeCoins, hasNeg = feeCoins.SafeSub(additionalFees); hasNeg {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, DefaultInsufficientFeeMsg+": %q, required additional fee: %q", feeCoins, additionalFees)
+	}
+	// Step 2: check if additional fees in nhash, that base fees and additional fees can be paid
+	// total fees in hash - gas limit * price per gas >= additional fees in hash
+	if containsDenom(additionalFees, minGasPriceForAdditionalFeeCalc.Denom) {
```

Comment:
cointainsDenom seems to be used once? It can be replaced with one of the two `sdk.Coin` functions that return the amount of a denom
```
// AmountOf returns the amount of a denom from coins
func (coins Coins) AmountOf(denom string) Int {
	mustValidateDenom(denom)
	return coins.AmountOfNoDenomValidation(denom)
}

// AmountOfNoDenomValidation returns the amount of a denom from coins
// without validating the denomination.
func (coins Coins) AmountOfNoDenomValidation(denom string) Int {
	switch len(coins) {
	case 0:
		return ZeroInt()

	case 1:
		coin := coins[0]
		if coin.Denom == denom {
			return coin.Amount
		}
		return ZeroInt()

	default:
		// Binary search the amount of coins remaining
		midIdx := len(coins) / 2 // 2:1, 3:1, 4:2
		coin := coins[midIdx]
		switch {
		case denom < coin.Denom:
			return coins[:midIdx].AmountOfNoDenomValidation(denom)
		case denom == coin.Denom:
			return coin.Amount
		default:
			return coins[midIdx+1:].AmountOfNoDenomValidation(denom)
		}
	}
}
```

Reply:
Deleted method, moved the `AmountOf(` call from that method to the if statement